cartesian_product doesn't work with filter_view

Reduced from [StackOverflow](https://stackoverflow.com/q/54029428/2069064):

```cpp
#include <range/v3/view/cartesian_product.hpp>
#include <range/v3/view/filter.hpp>

int main() {
    int data[] = {1};
    ranges::view::cartesian_product(
        data | ranges::view::filter([](int){ return true; }),
        data);
}
```
fails with a huge error, the top of which is:
```
In file included from /opt/compiler-explorer/libs/rangesv3/trunk/include/range/v3/view/cartesian_product.hpp:21,
                 from <source>:1:
/opt/compiler-explorer/libs/rangesv3/trunk/include/range/v3/range_concepts.hpp: In instantiation of 'class ranges::v3::cartesian_product_view<ranges::v3::remove_if_view<ranges::v3::iterator_range<int*, int*>, ranges::v3::logical_negate_<main()::<lambda(int)> > >, ranges::v3::iterator_range<int*, int*> >':
<source>:8:17:   required from here
/opt/compiler-explorer/libs/rangesv3/trunk/include/range/v3/range_concepts.hpp:78:50: error: no match for call to '(const ranges::v3::_begin_::fn) (const ranges::v3::remove_if_view<ranges::v3::iterator_range<int*, int*>, ranges::v3::logical_negate_<main()::<lambda(int)> > >&)'
   78 |                 using iterator_t = decltype(begin(std::declval<T &>()));
      |                                             ~~~~~^~~~~~~~~~~~~~~~~~~~~
In file included from /opt/compiler-explorer/libs/rangesv3/trunk/include/range/v3/view/cartesian_product.hpp:19,
                 from <source>:1:
/opt/compiler-explorer/libs/rangesv3/trunk/include/range/v3/begin_end.hpp:88:32: note: candidate: 'template<class R> constexpr decltype (ranges::v3::_begin_::fn::impl_(r, 42)) ranges::v3::_begin_::fn::operator()(R&) const'
   88 |                 constexpr auto operator()(R &r) const
      |                                ^~~~~~~~
```

--- a/include/range/v3/view/cartesian_product.hpp
+++ b/include/range/v3/view/cartesian_product.hpp
@@ -74,29 +74,39 @@ namespace ranges
             friend range_access;
             CONCEPT_ASSERT(sizeof...(Views) != 0);
             CONCEPT_ASSERT(meta::strict_and<ForwardView<Views>...>::value);
-            using CanConst = meta::strict_and<
-                Range<Views const>...>;
+            using CanConst = meta::and_<Range<Views const>...>;
             template<bool IsConst>
-            using CanSize = meta::strict_and<
+            using CanSize = meta::and_<
                 SizedRange<meta::if_c<IsConst, Views const, Views>>...>;
             template<bool IsConst>
-            using CanDistance = meta::strict_and<
-                CanSize<IsConst>,
+            using CanDistance2_ = meta::and_<
                 SizedSentinel<
                     iterator_t<meta::if_c<IsConst, Views const, Views>>,
                     iterator_t<meta::if_c<IsConst, Views const, Views>>>...>;
             template<bool IsConst>
-            using CanRandom = meta::strict_and<
-                CanDistance<IsConst>,
+            using CanDistance = meta::and_<
+                CanSize<IsConst>,
+                CanDistance2_<IsConst && CanSize<IsConst>::value>>;
+            template<bool IsConst>
+            using CanRandom2_ = meta::and_<
                 RandomAccessIterator<iterator_t<
                     meta::if_c<IsConst, Views const, Views>>>...>;
             template<bool IsConst>
-            using CanBidi = meta::strict_or<
+            using CanRandom = meta::and_<
+                CanDistance<IsConst>,
+                CanRandom2_<IsConst && CanDistance<IsConst>::value>>;
+            template<bool IsConst>
+            using CanBidi2_ = meta::and_<
+                BidirectionalIterator<iterator_t<
+                    meta::if_c<IsConst, Views const, Views>>>...>;
+            template<bool IsConst>
+            using Bounded = meta::and_<
+                BoundedRange<meta::if_c<IsConst, Views const, Views>>...>;
+            template<bool IsConst>
+            using CanBidi = meta::or_<
                 CanRandom<IsConst>,
-                meta::strict_and<
-                    BoundedRange<meta::if_c<IsConst, Views const, Views>>...,
-                    BidirectionalIterator<iterator_t<
-                        meta::if_c<IsConst, Views const, Views>>>...>>;
+                meta::and_<Bounded<IsConst>,
+                    CanBidi2_<IsConst && Bounded<IsConst>::value>>>;
 
             static constexpr auto my_cardinality =
                 detail::cartesian_product_cardinality<Views...>::value;
@@ -240,10 +250,6 @@ RANGES_DIAGNOSTIC_IGNORE_DIVIDE_BY_ZERO
                     i = first + static_cast<D>(n_mod);
                 }
 RANGES_DIAGNOSTIC_POP
-                void check_at_end_(meta::size_t<0>, bool = false)
-                {
-                    CONCEPT_ASSERT(sizeof...(Views) == 0);
-                }
                 void check_at_end_(meta::size_t<1>, bool at_end = false)
                 {
                     if(at_end)
@@ -333,26 +339,23 @@ RANGES_DIAGNOSTIC_POP
             {
                 return cursor<true>{begin_tag{}, *this};
             }
-            CONCEPT_REQUIRES(sizeof...(Views) == 0 || CanBidi<false>())
+            CONCEPT_REQUIRES(CanBidi<false>())
             cursor<false> end_cursor()
             {
-                using Tag = meta::if_c<sizeof...(Views) == 0, begin_tag, end_tag>;
-                return cursor<false>{Tag{}, *this};
+                return cursor<false>{end_tag{}, *this};
             }
-            CONCEPT_REQUIRES(sizeof...(Views) == 0 || CanBidi<true>())
+            CONCEPT_REQUIRES(CanBidi<true>())
             cursor<true> end_cursor() const
             {
-                using Tag = meta::if_c<sizeof...(Views) == 0, begin_tag, end_tag>;
-                return cursor<true>{Tag{}, *this};
+                return cursor<true>{end_tag{}, *this};
             }
-            CONCEPT_REQUIRES(sizeof...(Views) != 0 && !CanBidi<true>())
+            CONCEPT_REQUIRES(!CanBidi<true>())
             default_sentinel end_cursor() const
             {
                 return {};
             }
         public:
             cartesian_product_view() = default;
-            CONCEPT_REQUIRES(sizeof...(Views) > 0)
             explicit constexpr cartesian_product_view(Views... views)
               : views_{detail::move(views)...}
             {}

--- a/test/view/cartesian_product.cpp
+++ b/test/view/cartesian_product.cpp
@@ -22,6 +22,7 @@
 #include <range/v3/view/cartesian_product.hpp>
 #include <range/v3/view/chunk.hpp>
 #include <range/v3/view/empty.hpp>
+#include <range/v3/view/filter.hpp>
 #include <range/v3/view/iota.hpp>
 #include <range/v3/view/reverse.hpp>
 #include <range/v3/view/take_exactly.hpp>
@@ -272,6 +273,12 @@ int main()
     test_bug_820();
     test_bug_823();
     test_bug_919();
+
+    {
+        static constexpr int data[] = {1};
+        ranges::view::cartesian_product(data |
+            ranges::view::filter([](int) { return true; }));
+    }
 
     return test_result();
 }

